### PR TITLE
[datadogexporter] Use a singleton for sublayer calculation

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

- Use a single sublayer calculator per trace exporter. This reduces memory consumption at the expense of increased mutex contention. Ideally, we would use a sublayer calculator per worker with no need for mutexes but this is not currently possible. 

**Link to tracking Issue:**  aws-observability/aws-otel-collector#179

**Testing:** Unit tests were amended. Tested end to end and performance tested.

### CODE REVIEW

**Review by ericmustin — APPROVED** [review]
👍 lgtm, approving on behalf of datadog

File: exporter/datadogexporter/stats.go [hunk 1]
```diff
@@ -26,8 +26,7 @@ const (
 )
 
 // ComputeAPMStats calculates the stats that should be submitted to APM about a given trace
-func ComputeAPMStats(tracePayload *pb.TracePayload, calculator *stats.SublayerCalculator, pushTime int64) *stats.Payload {
-
+func ComputeAPMStats(tracePayload *pb.TracePayload, calculator *sublayerCalculator, pushTime int64) *stats.Payload {
```
Comment (bogdandrutu) [inline exporter/datadogexporter/stats.go:29]: Why public?

File: exporter/datadogexporter/traces_exporter.go [hunk 2]
```diff
@@ -32,11 +33,27 @@ import (
...
+// sublayerCalculator is thread safe wrapper of a sublayer
+// calculator. Each trace exporter has a single sublayer
+// calculator that is reused by each push
+type sublayerCalculator struct {
+	sc    *stats.SublayerCalculator
+	mutex sync.Mutex
+}
+
+// ComputeSublayers computes the sublayers of a trace
+func (s *sublayerCalculator) ComputeSublayers(trace pb.Trace) []stats.SublayerValue {
```
Comment (bogdandrutu) [inline exporter/datadogexporter/traces_exporter.go:45]: Why public?

File: exporter/datadogexporter/stats.go (hunk 1 above)
Comment (mx-psi) [reply exporter/datadogexporter/stats.go]: I was trying to maintain the visibility of existing functions that were already public. I have double checked with the person who wrote this code and these and other functions can be made private, so I have done so. 

I can also make these changes in a separate PR if need be.

File: exporter/datadogexporter/traces_exporter.go (hunk 2 above)
Comment (mx-psi) [reply exporter/datadogexporter/traces_exporter.go]: See reply in the other comment, marking as solved.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**